pkg/controller/mutators/core/controller_core.go

```diff
@@ -0,0 +1,314 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	opa "github.com/open-policy-agent/frameworks/constraint/pkg/client"
+	mutationsv1alpha1 "github.com/open-policy-agent/gatekeeper/apis/mutations/v1alpha1"
+	statusv1beta1 "github.com/open-policy-agent/gatekeeper/apis/status/v1beta1"
+	ctrlmutators "github.com/open-policy-agent/gatekeeper/pkg/controller/mutators"
+	"github.com/open-policy-agent/gatekeeper/pkg/controller/mutatorstatus"
+	"github.com/open-policy-agent/gatekeeper/pkg/logging"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+	"github.com/open-policy-agent/gatekeeper/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	"github.com/open-policy-agent/gatekeeper/pkg/watch"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apiTypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type Adder struct {
+	MutationSystem *mutation.System
+	Tracker        *readiness.Tracker
+	GetPod         func() (*corev1.Pod, error)
+	Kind           string
+	NewObj         func() client.Object
+	NewMutator     func(client.Object) (types.Mutator, error)
+}
+
+// Add creates a new Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (a *Adder) Add(mgr manager.Manager) error {
+	r := newReconciler(mgr, a.MutationSystem, a.Tracker, a.GetPod, a.Kind, a.NewObj, a.NewMutator)
+	return add(mgr, r)
+}
+
+func (a *Adder) InjectOpa(o *opa.Client) {}
+
+func (a *Adder) InjectWatchManager(w *watch.Manager) {}
+
+func (a *Adder) InjectControllerSwitch(cs *watch.ControllerSwitch) {}
+
+func (a *Adder) InjectTracker(t *readiness.Tracker) {
+	a.Tracker = t
+}
+
+func (a *Adder) InjectGetPod(getPod func() (*corev1.Pod, error)) {
+	a.GetPod = getPod
+}
+
+func (a *Adder) InjectMutationSystem(mutationSystem *mutation.System) {
+	a.MutationSystem = mutationSystem
+}
```

Comment:
These functions exist in both `controller_core.go` and both mutator controllers.  Are they needed here as well?  It seems like the `Adder` here will never be tasked with meeting the [`Injector`](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/controller/controller.go#L42) interface.  Disambiguating this usage from the _actual_ implementations of `Injector` interface might make this clearer.

Reply:
Yeah, these are superfluous, I can remove them